nemo/collections/asr/parts/k2/map_loss.py

```diff
@@ -214,84 +203,125 @@ def _intersect_calc_scores_mmi_pruned(
         )
 
-        # use_double_scores=True does matter
-        # since otherwise it sometimes makes rounding errors
-        num_tot_scores = num_lats.get_tot_scores(log_semiring=True, use_double_scores=True)
-        den_tot_scores = den_lats.get_tot_scores(log_semiring=True, use_double_scores=True)
+        num_tot_scores = num_lats.get_tot_scores(log_semiring=True, use_double_scores=False)
+        den_tot_scores = den_lats.get_tot_scores(log_semiring=True, use_double_scores=False)
 
         if return_lats:
             return num_tot_scores, den_tot_scores, num_lats, den_lats
         else:
             return num_tot_scores, den_tot_scores, None, None
 
-    def forward(
+    def _intersect_calc_scores(
         self,
-        log_probs: torch.Tensor,
-        targets: torch.Tensor,
-        input_lengths: torch.Tensor,
-        target_lengths: torch.Tensor,
-    ) -> torch.Tensor:
-        assert self.graph_compiler is not None
-        boosted = self.boost_coeff != 0.0
-        if self.blank != 0:
-            # rearrange log_probs to put blank at the first place
-            # and shift targets to emulate blank = 0
-            log_probs, targets = make_blank_first(self.blank, log_probs, targets)
-        supervisions, order = create_supervision(input_lengths)
-        order = order.long()
-        targets = targets[order]
-        target_lengths = target_lengths[order]
-
-        if log_probs.device != self.graph_compiler.device:
-            self.graph_compiler.to(log_probs.device)
-
-        num_graphs, den_graph = self.graph_compiler.compile(
-            targets + 1 if self.pad_fsavec else targets, target_lengths
-        )
+        emissions_graphs: 'k2.DenseFsaVec',
+        supervision_graphs: Tuple['k2.Fsa', 'k2.Fsa'],
+        supervisions: torch.Tensor,
+    ) -> Tuple[torch.Tensor, torch.Tensor]:
+        """Intersects emissions_graphs with supervision_graphs and calculates lattice scores.
+        This version assumes supervision_graphs to be a pair of the numerator and the denominator FSAs.
 
-        dense_fsa_vec = (
-            prep_padded_densefsavec(log_probs, supervisions)
-            if self.pad_fsavec
-            else k2.DenseFsaVec(log_probs, supervisions)
-        )
+        It can also calculate accuracy between the numerator and the denominator lattices to use it as additional loss.
 
-        num_tot_scores, den_tot_scores, num_lats, den_lats = self.intersect_calc_scores(
-            dense_fsa_vec, num_graphs, den_graph, boosted
+        Can be overridden.
+        """
+        boosted = self.boost_coeff != 0.0
+        num_tot_scores, den_tot_scores, num_lats, den_lats = self._intersect_calc_scores_impl(
+            emissions_graphs, supervision_graphs[0], supervision_graphs[1], boosted
         )
 
-        tot_scores = num_tot_scores - den_tot_scores
+        inverted_batch_order = invert_permutation(supervisions[:, 0].to(dtype=torch.long))
```

Comment:
Seems that `invert_permutation` not imported (undefined)

Reply:
fixed